### PR TITLE
CSV parser

### PR DESCRIPTION
**Pull request by DSDubit**
Utility which allows the caller to specify a format of expected CSV data, then parse a file which matches that  format in a type-safe way, including converting the input to common (int, bool, float, enum<T>) value types, and custom conversions (including e.g. value range constraints), then validating the entire document.

Deluxe(TM)

### CODE REVIEW

**Comment by DSDubit** [comment]
Use case from TOPPS:

- define some custom types which extend the basic CSV.Type, but with input pre-manipulation, value range limits, etc
- create a dictionary of field names to expected types
- parse a CSV file with this format
- optionally reject empty values (e.g. cardID), or support them as default(T) (e.g. player_name)

```c#

                private static CSV.Type cardValue = new CSV.Type<float>((str) =>
		{
			return float.Parse(str.Replace(" Million", string.Empty));
		}, allowEmptyValues: true);

		private static CSV.Type cardStat = new CSV.Types.Int(allowEmptyValues: true, minValue: 0, maxValue: 101);

		private static CSV.Type cardFoil = new CSV.Type<CardDefines.FoilType>((str) =>
		{
			return (CardDefines.FoilType)System.Enum.Parse(typeof(CardDefines.FoilType), "Foil_" + str);
		}, allowEmptyValues: true);

                ingestionModel = new Dictionary<string, CSV.Type>
		{
			// Generic
			{ CSVKey.card_id.ToString(), new CSV.Types.String() },
			{ CSVKey.type.ToString(), new CSV.Types.Enum<CardDefines.CardType>() },

			// Visual
			{ CSVKey.club.ToString(), new CSV.Types.String(allowEmptyValues: true) },

			// Player
			{ CSVKey.position.ToString(), new CSV.Types.Enum<CardDefines.Role>(allowEmptyValues: true) },
			{ CSVKey.player_name.ToString(), new CSV.Types.String(allowEmptyValues: true) },
			{ CSVKey.country_flag.ToString(), new CSV.Types.String(allowEmptyValues: true) },
			{ CSVKey.value.ToString(), cardValue },
			{ CSVKey.defence.ToString(), cardStat },
			{ CSVKey.attack.ToString(), cardStat },
			{ CSVKey.stars.ToString(), new CSV.Types.Int(allowEmptyValues: true, minValue: 0, maxValue: 5) },
			{ CSVKey.speed.ToString(), cardStat },
			{ CSVKey.tackle.ToString(), cardStat },
			{ CSVKey.power.ToString(), cardStat },
			{ CSVKey.shoot.ToString(), cardStat },
			{ CSVKey.skill.ToString(), cardStat },
			{ CSVKey.pass.ToString(), cardStat },
			{ CSVKey.foil.ToString(), cardFoil }
		};

                var doc = CSV.Parse(csvFile.text, new CSV.Format(ingestionModel), allowUnexpectedFields: false);
```

File: Utils/CSV.cs [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+						// Field isn't known
+						else
+						{
+							if (!allowUnexpectedFields)
+							{
+								throw new Exception("Unknown field: " + fieldName);
+							}
+							else
+							{
+								format.Add(fieldName.Trim(), defaultType);
```
Comment (FenrirWillow) [inline Utils/CSV.cs]: Is there a standard approach in DUCK in the way that default failover behaviour is communicated to it's user. For example I think that here we should log out a warning in the console, saying that this field name will default failover and print what that value would be.

**Comment by FenrirWillow** [comment]
I can see that there is an example provided in the PR description - can we add this to some sort of doc somewhere. Helps visualize how the API interacts and everything.

File: Utils/CSV.cs (hunk 1 above)
Comment (DSDubit) [reply Utils/CSV.cs]: Sounds good, I'll add a warning

**Comment by DSDubit** [comment]
Documentation sounds reasonable, but we currently have no provision in DUCK for readmes, do we? Where would I put something like this? A separate file, comment block in the class, or scattered comments at each point of use?

**Pull request closed without merging**